File: src/bluesky/core/interop.py
"""
Interoperability utilities for calling Ruby from Python.

This module provides helpers for Python scripts to execute Ruby code and exchange data.
"""

import json
import subprocess
import sys
from pathlib import Path
from typing import Any, Optional


class RubyExecutionError(Exception):
    """Raised when Ruby script execution fails."""

    pass


def run_ruby_script(
    script_path: str | Path,
    args: list[str] | None = None,
    input_data: dict[str, Any] | None = None,
    timeout: int = 300,
) -> dict[str, Any]:
    """
    Execute a Ruby script from Python.

    Args:
        script_path: Path to the Ruby script
        args: Command-line arguments for the script
        input_data: Dictionary to pass to Ruby via JSON (stdin)
        timeout: Execution timeout in seconds

    Returns:
        dict: Result data from Ruby script (via JSON stdout)

    Raises:
        RubyExecutionError: If execution fails
        FileNotFoundError: If Ruby or script not found

    Example:
        >>> result = run_ruby_script(
        ...     'create_model.rb',
        ...     input_data={'building_type': 'Office', 'floors': 3}
        ... )
        >>> print(result['model_path'])
    """

    script_path = Path(script_path)
    if not script_path.exists():
        raise FileNotFoundError(f"Ruby script not found: {script_path}")

    # Build command
    cmd = ["ruby", str(script_path)]
    if args:
        cmd.extend(args)

    # Prepare input data as JSON
    stdin_input = None
    if input_data:
        stdin_input = json.dumps(input_data)

    try:
        result = subprocess.run(
            cmd, input=stdin_input, capture_output=True, text=True, timeout=timeout, check=True
        )

        # Try to parse JSON output
        try:
            return json.loads(result.stdout)
        except json.JSONDecodeError:
            # If not JSON, return raw output
            return {"stdout": result.stdout, "stderr": result.stderr}

    except subprocess.CalledProcessError as e:
        raise RubyExecutionError(
            f"Ruby script failed with exit code {e.returncode}\n"
            f"Stdout: {e.stdout}\n"
            f"Stderr: {e.stderr}"
        )
    except subprocess.TimeoutExpired:
        raise RubyExecutionError(f"Ruby script timed out after {timeout} seconds")


def call_ruby_function(
    script_path: str | Path, function_name: str, **kwargs
) -> dict[str, Any]:
    """
    Call a specific Ruby function with arguments.

    This is a convenience wrapper around run_ruby_script that passes
    function name and arguments as JSON.

    Args:
        script_path: Path to Ruby script
        function_name: Name of Ruby function/method to call
        **kwargs: Arguments to pass to the Ruby function

    Returns:
        dict: Result from Ruby function

    Example:
        >>> result = call_ruby_function(
        ...     'model_ops.rb',
        ...     'create_necb_model',
        ...     climate_zone='6',
        ...     building_type='Office'
        ... )
    """

    input_data = {"function": function_name, "args": kwargs}

    return run_ruby_script(script_path, input_data=input_data)


def exchange_via_file(data: dict[str, Any], exchange_path: str | Path) -> None:
    """
    Write data to JSON file for Ruby to read.

    Args:
        data: Dictionary to write
        exchange_path: Path to exchange file

    Example:
        >>> exchange_via_file(
        ...     {'model_path': 'model.osm', 'simulate': True},
        ...     'exchange.json'
        ... )
    """

    exchange_path = Path(exchange_path)
    with open(exchange_path, "w") as f:
        json.dump(data, f, indent=2)


def read_exchange_file(exchange_path: str | Path) -> dict[str, Any]:
    """
    Read data from JSON exchange file written by Ruby.

    Args:
        exchange_path: Path to exchange file

    Returns:
        dict: Data from Ruby script

    Example:
        >>> result = read_exchange_file('results.json')
    """

    exchange_path = Path(exchange_path)
    with open(exchange_path) as f:
        return json.load(f)


def find_ruby_executable() -> Optional[str]:
    """
    Find Ruby executable in system PATH.

    Returns:
        str: Path to Ruby, or None if not found
    """

    import shutil

    return shutil.which("ruby")


def check_ruby_gem(gem_name: str) -> bool:
    """
    Check if a Ruby gem is installed.

    Args:
        gem_name: Name of the gem (e.g., 'openstudio-standards')

    Returns:
        bool: True if gem is installed

    Example:
        >>> if check_ruby_gem('openstudio-standards'):
        ...     print("Standards library available")
    """

    try:
        result = subprocess.run(
            ["gem", "list", "-i", gem_name], capture_output=True, text=True, timeout=10
        )
        return result.returncode == 0
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return False

File: src/bluesky/core/interop.rb
# frozen_string_literal: true

# Interoperability utilities for calling Python from Ruby.
#
# This module provides helpers for Ruby scripts to execute Python code and exchange data.

require 'json'
require 'open3'

module Bluesky
  module Interop
    # Error raised when Python execution fails
    class PythonExecutionError < StandardError; end

    # Execute a Python script from Ruby
    #
    # @param script_path [String] Path to the Python script
    # @param args [Array<String>] Command-line arguments
    # @param input_data [Hash] Data to pass to Python via JSON (stdin)
    # @param timeout [Integer] Execution timeout in seconds
    # @return [Hash] Result data from Python (via JSON stdout)
    #
    # @example
    #   result = Bluesky::Interop.run_python_script(
    #     'analyze.py',
    #     input_data: { model_path: 'model.osm', metrics: ['energy', 'cost'] }
    #   )
    #   puts result['annual_energy']
    def self.run_python_script(script_path, args: [], input_data: nil, timeout: 300)
      raise "Python script not found: #{script_path}" unless File.exist?(script_path)

      # Build command
      cmd = ['python3', script_path] + args

      # Prepare input data as JSON
      stdin_data = input_data ? JSON.generate(input_data) : nil

      # Execute with timeout
      stdout, stderr, status = Open3.capture3(*cmd, stdin_data: stdin_data, timeout: timeout)

      unless status.success?
        raise PythonExecutionError, "Python script failed with exit code #{status.exitstatus}\nStdout: #{stdout}\nStderr: #{stderr}"
      end

      # Try to parse JSON output
      begin
        JSON.parse(stdout)
      rescue JSON::ParserError
        # If not JSON, return raw output
        { 'stdout' => stdout, 'stderr' => stderr }
      end
    rescue Timeout::Error
      raise PythonExecutionError, "Python script timed out after #{timeout} seconds"
    end

    # Call a specific Python function with arguments
    #
    # @param script_path [String] Path to Python script
    # @param function_name [String] Name of Python function to call
    # @param kwargs [Hash] Keyword arguments for the function
    # @return [Hash] Result from Python function
    #
    # @example
    #   result = Bluesky::Interop.call_python_function(
    #     'analysis.py',
    #     'calculate_metrics',
    #     model_path: 'model.osm',
    #     include_hourly: true
    #   )
    def self.call_python_function(script_path, function_name, **kwargs)
      input_data = {
        'function' => function_name,
        'args' => kwargs
      }

      run_python_script(script_path, input_data: input_data)
    end

    # Write data to JSON file for Python to read
    #
    # @param data [Hash] Data to write
    # @param exchange_path [String] Path to exchange file
    #
    # @example
    #   Bluesky::Interop.exchange_via_file(
    #     { model_path: 'model.osm', results: results_hash },
    #     'exchange.json'
    #   )
    def self.exchange_via_file(data, exchange_path)
      File.write(exchange_path, JSON.pretty_generate(data))
    end

    # Read data from JSON exchange file written by Python
    #
    # @param exchange_path [String] Path to exchange file
    # @return [Hash] Data from Python script
    def self.read_exchange_file(exchange_path)
      JSON.parse(File.read(exchange_path))
    end

    # Find Python executable in system PATH
    #
    # @return [String, nil] Path to Python, or nil if not found
    def self.find_python_executable
      %w[python3 python].each do |cmd|
        path = `which #{cmd}`.strip
        return path unless path.empty?
      end
      nil
    end

    # Check if a Python package is installed
    #
    # @param package_name [String] Name of the package (e.g., 'openstudio')
    # @return [Boolean] True if package is installed
    #
    # @example
    #   if Bluesky::Interop.check_python_package('h2k_hpxml')
    #     puts "H2K-HPXML library available"
    #   end
    def self.check_python_package(package_name)
      cmd = ['python3', '-c', "import #{package_name}"]
      _, _, status = Open3.capture3(*cmd)
      status.success?
    rescue StandardError
      false
    end
  end
end
